sgm/inference/helpers.py

```diff
@@ -276,30 +313,73 @@ def do_img2img(
 
                 if offset_noise_level > 0.0:
                     noise = noise + offset_noise_level * append_dims(
                         torch.randn(z.shape[0], device=z.device), z.ndim
                     )
-                noised_z = z + noise * append_dims(sigma, z.ndim)
-                noised_z = noised_z / torch.sqrt(
-                    1.0 + sigmas[0] ** 2.0
-                )  # Note: hardcoded to DDPM-like scaling. need to generalize later.
+                if add_noise:
+                    noised_z = z + noise * append_dims(sigma, z.ndim).cuda()
+                    noised_z = noised_z / torch.sqrt(
+                        1.0 + sigmas[0] ** 2.0
+                    )  # Note: hardcoded to DDPM-like scaling. need to generalize later.
+                else:
+                    noised_z = z / torch.sqrt(1.0 + sigmas[0] ** 2.0)
 
                 def denoiser(x, sigma, c):
                     return model.denoiser(model.model, x, sigma, c)
 
-                samples_z = sampler(denoiser, noised_z, cond=c, uc=uc)
-                samples_x = model.decode_first_stage(samples_z)
-                samples = torch.clamp((samples_x + 1.0) / 2.0, min=0.0, max=1.0)
+                with SwapToDevice(model.denoiser, device):
+                    with SwapToDevice(model.model, device):
+                        samples_z = sampler(denoiser, noised_z, cond=c, uc=uc)
+
+                with SwapToDevice(model.first_stage_model, device):
+                    samples_x = model.decode_first_stage(samples_z)
+                    samples = torch.clamp((samples_x + 1.0) / 2.0, min=0.0, max=1.0)
 
                 if filter is not None:
                     samples = filter(samples)
 
                 if return_latents:
                     return samples, samples_z
                 return samples
+
+
+class SwapToDevice(object):
+    def __init__(
+        self,
+        model: Union[torch.nn.Module, torch.Tensor],
+        device: Union[torch.device, str],
+    ):
+        self.model = model
+        self.device = torch.device(device)
+        if isinstance(model, torch.Tensor):
+            self.original_device = model.device
+        else:
+            param = next(model.parameters(), None)
+            if param is not None:
+                self.original_device = param.device
+            else:
+                buf = next(model.buffers(), None)
+                if buf is not None:
+                    self.original_device = buf.device
+                else:
+                    # If device could not be found, turn this into a no-op
+                    self.original_device = self.device
+
+    def __enter__(self):
+        if self.device != self.original_device:
+            self.model.to(self.device)
+
+    def __exit__(self, *args):
+        if self.device != self.original_device:
+            self.model.to(self.original_device)
+            if torch.cuda.is_available():
+                torch.cuda.empty_cache()
```

Comment:
This assumes `cuda` is involved (see #49).

Reply:
> This assumes `cuda` is involved (see #49).

It needs to be expanded to cover other device types, but it should at least not fail, right?

Reply:
It won't fail, but it won't clear stuff either (there's a separate function for `mps` cleaning, etc.)